src/NUnitCommon/nunit.agent.core/Drivers/NUnitFrameworkDriver.cs

```diff
@@ -152,5 +155,73 @@ public string Load(string testAssemblyPath, IDictionary<string, object> settings
         /// <param name="filter">A filter indicating which tests to include</param>
         /// <returns>An Xml string representing the tests</returns>
         public string Explore(string filter) => _api.Explore(filter);
+
+        /// <summary>
+        /// Nested class used to intercept progress reports received from the test
+        /// framework and resend them to the actual listener, normally located in
+        /// the runner that is using the engine.
+        /// </summary>
+        /// <remarks>
+        /// This class is absolutely needed when the 2018 NUnit API is used under
+        /// the .NET Framework using Windows Remoting as a communicatioion protocol.
+        /// In particular, the MarshalByRef object implementing the listener must 
+        /// be convertible to ITestEventListener via the IConvertible interface.
+        /// 
+        /// In other cases, the interceptor is not essential, but we use it anyway
+        /// for several reasons:
+        /// 
+        /// 1. We have no control over the implementation of runners using the engine.
+        ///    They may not implement IConvertible and may not even derive from 
+        ///    MarshaByRefObject.
+        /// 
+        /// 2. The interceptor provides a point of control for checking what events 
+        ///    are received from the framework and for possible future modifications to
+        ///    the events before they are forwarded.
+        /// </remarks>
+        public class EventInterceptor : MarshalByRefObject, ITestEventListener, IConvertible
+        {
+            private ITestEventListener _listener;
+
+            public EventInterceptor(ITestEventListener listener)
+            {
+                _listener = listener;
+            }
+
+            #region ITestEventListener and IConvertible Implementations
+
+            void ITestEventListener.OnTestEvent(string report)
+            {
+                _listener.OnTestEvent(report);
+            }
+
+
+            // Conversion to ITestEventListener is the only one that makes sense
+            object IConvertible.ToType(Type conversionType, IFormatProvider? provider) =>
+                conversionType == typeof(ITestEventListener) ? this : InvalidCast(conversionType);
```

Comment:
This code looks odd, since it "converts" the object by returning it as an object, but it turns out that this is what you have to do in this situation. And, of course, you have to implement all the IConvertible methods even though none of them do anything but throw. Hopefully, we'll stop using Windows Remoting at some point and be able to get rid of all this.